keras_nlp/models/gpt2/gpt2_causal_lm.py

```diff
@@ -200,17 +213,138 @@ def backbone_cls(cls):
     def preprocessor_cls(cls):
         return GPT2CausalLMPreprocessor
 
-    def _get_token_probability(self, prompt, mask):
+    def call_with_cache(self, inputs, cache, cache_index=None):
+        """Forward pass of `GPT2CausalLM` with cache.
+
+        The difference between `call_with_cache` and normal `__call__` is in
+        this method, a `cache` arg is set, and the inputs is of
+        `sequence_length=1`. By cachine the previous key/value in multi-head
+        attention, we avoid recomputing the outputs of seen tokens.
+
+        Args:
+            inputs: a dict of key `token_ids` and `padding_mask`, the same
+                format as `GPT2Backbone` inputs.
+            cache: a dense float Tensor, the cache of key and value.
+            cache_index: int, or int Tensor, defaults to None. If set, it
+                represents the index of current inputs in the whole sequence.
+
+        Returns:
+            x: a dense float Tensor, the next token logits of `inputs`.
+            cache: a dense float Tensor, the updated cache.
+        """
+        token_ids = inputs["token_ids"]
+        padding_mask = inputs["padding_mask"]
+        token_embedding = self.token_embeddings(token_ids)
+        if cache_index is None:
+            position_embedding = self.position_embeddings(token_embedding)
+        else:
+            position_embedding = self.position_embeddings.position_embeddings[
+                cache_index, :
+            ]
+        x = token_embedding + position_embedding
```

Comment:
I would actually give a name to the `add` layer from the base class, and use that layer here. To keep the logic of the forward pass as close as possible.

Reply:
sounds good, refactored.